common/flatpak-run.c

```diff
@@ -1691,6 +1692,180 @@ flatpak_run_allocate_id (int *lock_fd_out)
+                          DConfClient *client,
+                          const char  *dir)
+{
+  g_auto(GStrv) keys = NULL;
+  int i;
+
+  keys = dconf_client_list (client, dir, NULL);
+  for (i = 0; keys[i]; i++)
+    {
+      g_autofree char *k = g_strconcat (dir, keys[i], NULL);
+      if (dconf_is_dir (k, NULL))
+        add_dconf_dir_to_keyfile (keyfile, client, k);
+      else if (dconf_is_key (k, NULL))
+        add_dconf_key_to_keyfile (keyfile, client, k);
+    }
+}
+
+static void
+add_dconf_locks_to_list (GString     *s,
+                         DConfClient *client,
+                         const char  *dir)
+{
+  g_auto(GStrv) locks = NULL;
+  int i;
+
+  locks = dconf_client_list_locks (client, dir, NULL);
+  for (i = 0; locks[i]; i++)
+    {
+      g_string_append (s, locks[i]);
+      g_string_append_c (s, '\n');
+    }
+}
+
+static char *
+dconf_path_for_app_id (const char *app_id)
+{
+  GString *s;
+  const char *p;
+
+  s = g_string_new ("");
+
+  g_string_append_c (s, '/');
+  for (p = app_id; *p; p++)
+    {
+      if (*p == '.')
+        g_string_append_c (s, '/');
+      else
+        g_string_append_c (s, *p);
+    }
+  g_string_append_c (s, '/');
+
+  return g_string_free (s, FALSE);
+}
+
+static void
+get_dconf_data (const char  *app_id,
+                const char **settings,
+                char **defaults,
+                gsize *defaults_size,
+                char **locks,
+                gsize *locks_size)
+{
+  DConfClient *client = NULL;
+  g_autoptr(GKeyFile) keyfile = NULL;
+  g_autofree char *path = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GString) contents = NULL;
+  g_autofree char *prefix = NULL;
+  g_autofree char *dir = NULL;
+
+  prefix = dconf_path_for_app_id (app_id);
+
+  client = dconf_client_new ();
+  keyfile = g_key_file_new ();
+
+  contents = g_string_new ("");
+
+  g_debug ("Add defaults in dir %s", prefix);
+  add_dconf_dir_to_keyfile (keyfile, client, prefix);
+
+  g_debug ("Add locks in dir %s", prefix);
+  add_dconf_locks_to_list (contents, client, prefix);
+
+  if (settings)
+    {
+      int i;
+      for (i = 0; settings[i]; i++)
+        {
+          if (dconf_is_dir (settings[i], NULL))
+            {
+              g_debug ("Add defaults in dir %s", settings[i]);
+              add_dconf_dir_to_keyfile (keyfile, client, settings[i]);
+
+              g_debug ("Add locks in dir %s", settings[i]);
+              add_dconf_locks_to_list (contents, client, settings[i]);
+            }
+          else if (dconf_is_key (settings[i], NULL))
+            {
+              g_debug ("Add individual key %s", settings[i]);
+              add_dconf_key_to_keyfile (keyfile, client, settings[i]);
+            }
+          else
+            {
+              g_warning ("Ignoring settings path '%s': neither dir nor key", settings[i]);
+            }
+        }
+    }
+
+  *defaults = g_key_file_to_data (keyfile, defaults_size, NULL);
+  *locks = g_strdup (contents->str);
```

Comment:
Please use `g_string_free (g_steal_pointer (&contents), FALSE)` here to avoid an extra allocation.